Set notifications m_tip_block in LoadChainTip()

Sjors opened the pull request:
Ensure KernelNotifications `m_tip_block` is set even if no new block arrives.

Suggested in https://github.com/bitcoin/bitcoin/pull/31297#issuecomment-2486457573

l0rinc commented on src/validation.cpp:
```diff
@@ -4706,6 +4706,11 @@ bool Chainstate::LoadChainTip()
+
+    // Ensure KernelNotifications m_tip_block is set even if no new block arrives.
+    // Ignoring return value for now.
+    (void)m_chainman.GetNotifications().blockTip(GetSynchronizationState(/*init=*/true, m_chainman.m_blockman.m_blockfiles_indexed), *pindex);
```
nit: at this point it may make more sense to just remove the `[[nodiscard]]` from `blockTip` instead

Sjors replied:
Maybe, or we could do something with the return value. I'm not familiar enough with init code and kernel notifications to have a strong opinion on this. cc @ryanofsky, @TheCharlatan

ryanofsky replied:
Just ignoring the return value is probably the best thing for now since this function and its callers return bool. After #29700 it could more easily bubble up the interrupted status. The kernel API is designed for a better world where operations can be more easily interrupted but not everything is interruptible not now.

Sjors commented:
I updated the `waitTipChanged()` and `m_tip_block` doc.

@maflcko which if-guard do you mean?

ryanofsky commented:
> @maflcko which if-guard do you mean?

I think probably this one:

```c++
    // Wait for genesis block to be processed
    if (WITH_LOCK(chainman.GetMutex(), return chainman.ActiveTip() == nullptr)) {
```

Removing the `if()` might help verify that `m_tip_block` is set reliably

Sjors commented:
I dropped the if guard.

> ensuring that any interfaces will only be spun up after that point in init

It could, but I'm not sure if it's worth putting such a restriction on the interfaces.

Similarly we have `startupnotify` which happens at the end of init. If we also ensure that the interfaces are ready to listen _before_ this notification, then at least from an application launched with systemd (or similiar) we could expect them to wait for this signal before trying to connect.

If on the other hand an external application is started at the time, it would just have to keep trying to connect until we listen. Such a race would exist even if we opened the interface a bit earlier in the init process, though it would be shorter.

Sjors commented:
Several reindex related tests are broken, investigating...

> After https://github.com/bitcoin/bitcoin/pull/31283 it might make sense to drop the `waitTipChanged` method entirely too.

I think it's generically useful though. There are various types of applications that need to know when a new block is added to the chain.

mzumsande commented on src/validation.cpp:
```diff
@@ -4706,6 +4706,11 @@ bool Chainstate::LoadChainTip()
+
+    // Ensure KernelNotifications m_tip_block is set even if no new block arrives.
+    // Ignoring return value for now.
+    (void)m_chainman.GetNotifications().blockTip(GetSynchronizationState(/*init=*/true, m_chainman.m_blockman.m_blockfiles_indexed), *pindex);
```
`LoadChainTip()` is called in a loop for all chainstates in `CompleteChainstateInitialization()` so that it would be called twice, for different blocks, in the case of an ongoing `AssumeUtxo` sync.
This seems problematic to me - how would subscribers to `blockTip` be supposed to deal with that? I think that the background chainstate shouldn't send a notification, as it it is done in `ActivateBestChain()`.

ryanofsky replied:
> `LoadChainTip()` is called in a loop for all chainstates in `CompleteChainstateInitialization()` so that `blockTip` would be called twice, for different blocks, in the case of an ongoing `AssumeUtxo` sync. This seems problematic to me - how would subscribers to `blockTip` be supposed to deal with that? I think that the background chainstate shouldn't send a notification, as it it is done in `ActivateBestChain()`.

Sjors commented on src/init.cpp:
```diff
@@ -1792,7 +1792,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     });
 
     // Wait for genesis block to be processed
```
I drafted a comment, but I think it's wrong so I haven't pushed it yet:

```cpp
    /* Wait for genesis block to be processed. Typically kernel_notifications.m_tip_block
     * has already been set, except in three cases where the genesis block is
     * being connected by the initload thread:
     *
     * 1. first startup with an empty datadir
     * 2. reindex
     * 3. reindex-chainstate
     */
```

I don't think the `initload` thread ever sets `kernel_notifications.m_tip_block`. The thread only deals with indexes derived from `BaseIndex`, which doesn't include `CBlockIndex`.

I'm still confused under what circumstance it would actually be unset.

Sjors replied:
It's clearly doing something though, because the following breaks several functional tests:

```patch
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -1792,13 +1792,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     });
 
     // Wait for genesis block to be processed
     {
         WAIT_LOCK(kernel_notifications.m_tip_block_mutex, lock);
-        kernel_notifications.m_tip_block_cv.wait(lock, [&]() EXCLUSIVE_LOCKS_REQUIRED(kernel_notifications.m_tip_block_mutex) {
-            return !kernel_notifications.m_tip_block.IsNull() || ShutdownRequested(node);
-        });
+        Assert(!kernel_notifications.m_tip_block.IsNull());
+        // kernel_notifications.m_tip_block_cv.wait(lock, [&]() EXCLUSIVE_LOCKS_REQUIRED(kernel_notifications.m_tip_block_mutex) {
+        //     return !kernel_notifications.m_tip_block.IsNull() || ShutdownRequested(node);
+        // });
     }
```

Sjors replied:
With some manual testing I indeed get all three conditions from your comment to hit that assert. I'm just confused why.

Sjors replied:
Outside the context of a snapshot, we only call `LoadChainTip()` if `if (!is_coinsview_empty(chainstate)) `:

https://github.com/bitcoin/bitcoin/blob/2638fdb4f934be96b7c798dbac38ea5ab8a6374a/src/node/chainstate.cpp#L147-L150

That explains why these three scenarios are different.

Still looking into where it _is_ set.

Sjors replied:
In the case of fresh data dir or reindex, it's `ActivateBestChain()` that calls `blockTip()` once it sets the genesis block to the tip.

This indeed happens in the `initload` threads, which seems to keep doing this throughout the reindex process. Once all previously stored blocks are done, it exits and the `msghand` thread takes over.

Sjors replied:
Specifically it's `ImportBlocks()` that does this, even if no blocks are passed to it. 

> which seems to keep doing this throughout the reindex process.

In fact it does it _after_ reindex is finished. `ActivateBestChain` connects blocks one by one until it has nothing left to do so that makes sense.

Sjors replied:
Took the comment and expanded it a bit.

mzumsande replied:
> In fact it does it after reindex is finished. ActivateBestChain connects blocks one by one until it has nothing left to do so that makes sense.

Just as an aside: ABC is also called [during](https://github.com/bitcoin/bitcoin/blob/2638fdb4f934be96b7c798dbac38ea5ab8a6374a/src/validation.cpp#L5149)  reindexing when encountering the genesis block. That's why we have this slightly weird behavior with a full `-reindex` that normally first reindexes all blocks and then builds the chainstate, but if the user interrupts it during reindexing and  restarts the node without `-reindex`, it will pick up by first connecting all blocks of the previously indexed block files, then reindex the remaining block files and finally connect the blocks from the second batch.

But that's not really relevant to this PR, the comment seems fine now!

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31346#discussion_r1856824817

> it will pick up by first connecting all blocks of the previously indexed block files

I don't think I understand what the slightly weird behavior is, or maybe what is meant by connecting all blocks. I'd expect if node is started without -reindex, `LoadChainTip` should connect all the blocks that were previously connected, and this would just avoid repeating work and not be weird.

mzumsande replied:
The "natural" IBD order is to first download all headers, add them to the block tree database, and then download all blocks / connect them when possible (with "connect" I mean `ActivateBestChain()` / `ConnectBlock()`).
During `reindex` we normally do it similarly: Go through all blockfiles, accept block headers to the db, don't connect any block except Genesis, and only in the end call `ActivateBestChain` to connect blocks / rebuild the chainstate (the reindex-chainstate part).

In the case of interrupt / restart, this order is mixed up. Naively I would expect to load the previously saved state (with only Genesis connected), continue to walk through the block files until all headers are loaded, and finally call `ActivateBestChain` to connect all blocks.

But in reality, we will call already `ActivateBestChain` early, when the best header is somewhere mid-chain, connect a first batch of blocks to the chainstate, then reindex the remaining block files, and then connect a second batch.

When I called this "weird" I did so because it mixes up the normal order, but I didn't know of a downside. However, it has a negative impact that @l0rinc pointed out to me (and I believe is working to fix): When we start connecting blocks without having all headers in our block tree db, we don't apply `assumevalid` (because our best header is not above the minchainwork threshold), and are therefore connecting blocks much slower while reindexing the chainstate.

ryanofsky replied:
Thanks, so is the scenario specifically the case when a node started with `-reindex` is stopped *before* the final ActivateBestChain call? If it is stopped *during* the final ActivateBestChain call it seems like it would be picking up work where it left off when ActivateBestChain was called again, due to `LoadChainTip` setting the previous tip.

> we will call already `ActivateBestChain` early, when the best header is somewhere mid-chain, connect a first batch of blocks to the chainstate, then reindex the remaining block files, and then connect a second batch.

This is referring to the [ActivateBestChain](https://github.com/bitcoin/bitcoin/blob/2638fdb4f934be96b7c798dbac38ea5ab8a6374a/src/validation.cpp#L5149) call in LoadExternalBlockFile call you linked to earlier? And removing this call would fix the weirdness and assumevalid issue? It does seem like it would be nice to remove this call. I'm not sure what the point of connecting the genesis block early but not connecting any other blocks is.

mzumsande replied:
> Thanks, so is the scenario specifically the case when a node started with `-reindex` is stopped _before_ the final ActivateBestChain call?

Yes.

>This is referring to the [ActivateBestChain](https://github.com/bitcoin/bitcoin/blob/2638fdb4f934be96b7c798dbac38ea5ab8a6374a/src/validation.cpp#L5149) call in LoadExternalBlockFile call you linked to earlier?

Yes!

>And removing this call would fix the weirdness and assumevalid issue? It does seem like it would be nice to remove this call. I'm not sure what the point of connecting the genesis block early but not connecting any other blocks is.

I don't think we'd want to remove it completely, because in case of a `-reindex` it allows `init` to complete normal startup without having to wait for the import thread to finish reindexing block files (which can take a long time for mainnet). I think that @l0rinc is looking into possible fixes, but without having tested anything my first idea was to only call `ActivateBestChain` there it if the chain is empty, so it only doesn't get called in the restart scenario where we already loaded Genesis during `LoadChainTip`.

l0rinc replied:
This is in my backlog, but I would be very happy if someone who already understands this fixes it instead :)
I just noticed it during Core Dev that my perf logs are full of script validations when trying out different scenarios and cancelling a reindex.

mzumsande replied:
> This is in my backlog, but I would be very happy if someone who already understands this fixes it instead :)

Ok thanks, I'll try out my suggestion from above and will open a PR if it's as straightforward as I think.

mzumsande replied:
#31439

ryanofsky commented on src/init.cpp:
```diff
@@ -1768,7 +1768,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     node.background_init_thread = std::thread(&util::TraceThread, "initload", [=, &chainman, &args, &node] {
         ScheduleBatchPriority();
-        // Import blocks
+        // Import blocks and ActivateBestChain()
```
s

ryanofsky replied:
>

ryanofsky reviewed (approved):
Code review ACK be78eaa9831be06bd94438d1f10ec8879b5c756d. Having `LoadChainTip` call `blockTip` makes a lot of sense and removes a confusing special case here and in followup PRs (#31297, #31283). The new comments are also really nice IMO.

mzumsande commented on src/node/blockstorage.cpp:
```diff
@@ -1206,7 +1206,7 @@ void ImportBlocks(ChainstateManager& chainman, std::span<const fs::path> import_
     ImportingNow imp{chainman.m_blockman.m_importing};
 
-    // -reindex
+    // Empty datadir or -reindex
```
This comment change not correct and should just be dropped: `m_blockfiles_indexed` is `true` (its default value) in case of an empty datadir, `false` in case of a reindex.

Sjors replied:
Ok, then I'm confused about what causes case (1):  https://github.com/bitcoin/bitcoin/pull/31346#discussion_r1855741271

mzumsande replied:
With an empty datadir, we don't have the genesis connected at this point. But we don't enter the reindex if clause [here](https://github.com/bitcoin/bitcoin/blob/2eccb8bc5e2245e42e05b8df3d9aa264b5f21a9a/src/node/blockstorage.cpp#L1210C16-L1210C17) (the comment only refers to that), we just skip this and the loadblock block and go to the [ActivateBestChain](https://github.com/bitcoin/bitcoin/blob/2eccb8bc5e2245e42e05b8df3d9aa264b5f21a9a/src/node/blockstorage.cpp#L1261) call below.

mzumsande reviewed (commented):
Looks good to me, just a request to change one comment, will A CK after.

Sjors commented:
Rebased (just in case) and dropped dropped comment: https://github.com/bitcoin/bitcoin/pull/31346#discussion_r1870315593

Sjors committed:
Set notifications m_tip_block in LoadChainTip()

Ensure KernelNotifications m_tip_block is set even if no new block arrives.

Additionally, have node init always wait for this to happen.

ryanofsky reviewed (approved):
Code review ACK 37946c0aafeebc1585f1316fb05f252f7fb51e91, fixing comment bug caught by @mzumsande in https://github.com/bitcoin/bitcoin/pull/31346#discussion_r1870315593 in another really helpful clarification

mzumsande reviewed (commented):
Code Review ACK 37946c0aafeebc1585f1316fb05f252f7fb51e91

sedited reviewed (approved):
ACK 37946c0aafeebc1585f1316fb05f252f7fb51e91

Sjors commented:
If you enjoyed this PR, you may like #31325.
